### PR TITLE
feat(action): show CI url in pullrequest body

### LINKED ISSUE #1717: Feature Request: Infer the GitHub action that created the GitHub Pull Request
### Is your feature request related to a problem?

Track what GitHub workflow created the GitHub Pull Request automatically.

### Solution you'd like

Infer the GitHub workflow URL when using the target `github/pullrequest` automatically and likely with an option to opt-out:

```yaml
actions:
  default:
    kind: github/pullrequest
    spec:
      inferGithubWorkflow: false
```

Will then include in the description of the Pull Request the URL to the gitHub workflow that triggered the udpatecli config.

### Alternatives you've considered

I've been using an env variable in conjunction with the GitHub action env generation

See https://github.com/elastic/apm-pipeline-library/blob/9c712c2558ef07309c4ddf3cf02d20a1cb2b1b3a/updatecli/updatecli.d/bump-aliases.yml#L14 and https://github.com/elastic/apm-pipeline-library/blob/9c712c2558ef07309c4ddf3cf02d20a1cb2b1b3a/.github/workflows/bump-aliases.yml#L12-L13

But I'd like to get that for free.

### Anything else?

`GITHUB_REPOSITORY`, `GITHUB_RUN_ID` and `GITHUB_SERVER_URL` are some of the required env variables that are available as far as I see in https://docs.github.com/en/actions/learn-github-actions/variables#default-environment-variables

https://github.com/elastic/apm-pipeline-library/pull/2398

image

### PR DESCRIPTION
**Pull request by olblak**
Fix #1717 

## Test

Well I don't see how to test this feature without taking the risk of conflicting with the GitHub Action environment used to test Updatecli so I am just doing manual testing at this stage.

We must use an Updatecli that need to change something

```shell
export GITHUB_ACTION="true" 
export GITHUB_SERVER_URL="https://github.com" 
export GITHUB_REPOSITORY="foo/bar" 
export GITHUB_RUN_ID="1658821493" 

./bin/updatecli diff --config updatecli/updatecli.d/venom.yaml --debug
```

Please note ` <a href="https://github.com/foo/bar/actions/runs/1658821493">GitHub Action pipeline link</a>` in the console output

<details><summary>Console Output</summary>

```
DEBUG: The expected action would have the following information:
	|	
	|	##Title:
	|	ci: bump Venom version to v1.0.1
	|	
	|	
	|	##Report:
	|	
	|	<Action id="1c70af8de95bd7989d99731b82355dfdd0fbd5981ebfef645550823023b51711">
	|	    <h3>ci: bump Venom version</h3>
	|	    <details id="dc3a5f47e02ede555033ee1e6cb765e5e88ea60e3352a90edb40cb190d142911">
	|	        <summary>ci: update Venom version to v1.0.1</summary>
	|	        <p>1 file(s) updated with &#34;VENOM_VERSION: v1.0.1&#34;:&#xA;&#x9;* .github/workflows/go.yaml&#xA;</p>
	|	        <details>
	|	            <summary>v1.0.1</summary>
	|	            <pre>&#xA;Release published on the 2021-12-13 16:52:12 +0000 UTC at the url https://github.com/ovh/venom/releases/tag/v1.0.1&#xA;&#xA;# Venom v1.0.1&#xD;&#xA;&#xD;&#xA;## What&#39;s Changed&#xD;&#xA;* fix: handle json number is assertions (#460) https://github.com/ovh/venom/pull/464&#xD;&#xA;&#xD;&#xA;**Full Changelog**: https://github.com/ovh/venom/compare/v1.0.0...v1.0.1</pre>
	|	        </details>
	|	    </details>
	|	    <a href="https://github.com/foo/bar/actions/runs/1658821493">GitHub Action pipeline link</a>
	|	</Action>
	|	
	|	=====
	|	

=============================

```

</details>

This pullrequest introduces in a new action setting `pipelineurl` such as 

```
actions:
    default:
        title: 'ci: bump Venom version to {{ source "latestVersion" }}'
        kind: github/pullrequest
        pipelineurl: true
        spec:
            automerge: true
            labels:
                - chore
                - skip-changelog
        scmid: default

```

### Potential improvement

- [x] I need to gate this feature behind a setting at the action level

### CODE REVIEW

**Review by dduportal — APPROVED** [review]
LGTM. I've added a few nitpicks, you're free to select them or not.


The last one (duplication of variables) is NOT a blocker. IF you want me to send a PR,n that will be a good exercise for me to put my hand dirty in `updatecli` again (haven't done it since a long time)

File: pkg/core/reports/action.go [hunk 1]
```diff
@@ -94,3 +102,55 @@ func (a Action) ToActionsString() string {
 
 	return string(output[:])
 }
+
+// UpdatePipelineURL analyze the local environment to guess if Updatecli is executed from a CI pipeline
+func (a *Action) UpdatePipelineURL() {
+
+	// isGitHubActionWorkflow check if the current execution is running from a GitHub Action
```
[the remark this answers is not in the document]
Comment (olblak) [reply pkg/core/reports/action.go:109]: Good point

File: pkg/core/reports/action.go [hunk 2]
```diff
@@ -94,3 +102,55 @@ func (a Action) ToActionsString() string {
...
+			logrus.Debugln("Jenkins build detected")
+			return true
+		}
+		return false
+	}
+
+	// isGitLabCI check if the current execution is running from a GitLab CI pipeline
+	isGitLabCI := func() bool {
+		if os.Getenv("CI_SERVER_URL") != "" &&
+			os.Getenv("CI_JOB_URL") != "" {
```
[the remark this answers is not in the document]
Comment (olblak) [reply pkg/core/reports/action.go:135]: This is a very very good point, I missed that variable name

File: pkg/core/reports/action.go [hunk 3]
```diff
@@ -94,3 +102,55 @@ func (a Action) ToActionsString() string {
...
+			logrus.Debugln("GitLab CI pipeline detected")
+			return true
+		}
+		return false
+	}
+
+	if isGitHubActionWorkflow() {
+		a.PipelineUrl = &PipelineURL{}
+		a.PipelineUrl.Name = "GitHub Action pipeline link"
+		a.PipelineUrl.URL = fmt.Sprintf(os.Getenv("GITHUB_SERVER_URL")+"/%s/actions/runs/%s", os.Getenv("GITHUB_REPOSITORY"), os.Getenv("GITHUB_RUN_ID"))
```
[the remark this answers is not in the document]
Comment (olblak) [reply pkg/core/reports/action.go:147]: That sounds great to me.

> I don't mind proposing a change after this PR is merged, or as a "PR to your PR",

Whatever works for you, maybe a follow up PR is easier to handle

[the remark this answers is not in the document]
Comment (dduportal) [reply pkg/core/reports/action.go:147]: Let's go for a subsequent PR then!

**Comment by dduportal** [comment]
I wonder if the `pipelineurl` shouldn't be `true` by default (e.g. opt-out) as it would not break any existing use case.

Eventually, it could be under the experimental flag  (e.g. enabled by default if `--experimental`) for early testing, and then in 1-2 version enabled by default.

WDYT @olblak ?

**Comment by olblak** [comment]
> I wonder if the pipelineurl shouldn't be true by default (e.g. opt-out) as it would not break any existing use case.

For me `pipelineurl` set to `false` means the current behavior which is not showing the pipeline url.
While set to `true` means we want to see the URL.

In long term I think having this opt-out by default would be better so I don't think the experimental flag is needed here as we don't the current behavior

**Comment by dduportal** [comment]
> > I wonder if the pipelineurl shouldn't be true by default (e.g. opt-out) as it would not break any existing use case.
> 
> For me `pipelineurl` set to `false` means the current behavior which is not showing the pipeline url. While set to `true` means we want to see the URL.
> 
> In long term I think having this opt-out by default would be better so I don't think the experimental flag is needed here as we don't the current behavior

I'm personally in favor of setting it to `true` by default, e.g. "if you use the new updatecli version, then you will have your PR with the link added by default"

**Review by dduportal — APPROVED** [review]
Approving, whatever default is chosen

**Comment by olblak** [comment]
> I'm personally in favor of setting it to true by default, e.g. "if you use the new updatecli version, then you will have your PR with the link added by default"

I was taking a more defensive approach, even thought I would also prefer to display the url by default

I am going to change the setting `pipelineurl` to `disablepipelineurl`

**Review by dduportal — APPROVED** [review]

File: pkg/core/reports/action.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (dduportal) [reply pkg/core/reports/action.go:147]: https://github.com/updatecli/updatecli/pull/1740

**Review by v1v — COMMENTED** [review]
💯 Thanks so much!!!

**Pull request merged**